WalletWasabi.Fluent.Generators/UIContextGenerator.cs

```diff
@@ -0,0 +1,313 @@
+		{
+			var (node, model) = (item.Node, item.SemanticModel);
+			if (node is not ClassDeclarationSyntax classDeclaration)
+			{
+				continue;
+			}
+
+			if (model.GetDeclaredSymbol(classDeclaration) is not INamedTypeSymbol classSymbol)
+			{
+				continue;
+			}
+
+			var ctors = GenerateConstructors(context, classDeclaration, model, classSymbol).ToList();
+
+			result.AddRange(ctors);
+		}
+
+		return result;
+	}
+
+	private static IEnumerable<ConstructorDeclarationSyntax> GenerateConstructors(SourceProductionContext context, ClassDeclarationSyntax classDeclaration, SemanticModel semanticModel, INamedTypeSymbol classSymbol)
+	{
+		var fileName = classDeclaration.Identifier.ValueText + UiContextAnalyzer.UiContextFileSuffix;
+
+		var className = classSymbol.Name;
+		var namespaceName = classSymbol.ContainingNamespace.ToDisplayString();
+
+		var ctors =
+			classDeclaration.ChildNodes()
+							.OfType<ConstructorDeclarationSyntax>()
+							.ToList();
+
+		foreach (var ctor in ctors)
+		{
+			if (!classDeclaration.GetUiContextReferences(semanticModel).Any())
+			{
+				if (!classDeclaration.IsAbstractClass(semanticModel) && ctor.IsPublic())
+				{
+					yield return ctor;
+				}
+			}
+			else
+			{
+				var ctorArgs =
+					  ctor.ParameterList.Parameters
+						  .Select(x => x.Identifier.ValueText)
+						  .ToList();
+
+				var hasCtorArgs = ctorArgs.Any();
+				var ctorArgsString = string.Join(",", ctorArgs);
+				var ctorString =
+					hasCtorArgs
+					? $": this({ctorArgsString})"
+					: "";
+
+				var parameterUsings =
+					ctor.ParameterList.Parameters
+									  .Where(p => p.Type is not null)
+									  .Select(p => semanticModel.GetTypeInfo(p.Type))
+									  .Where(t => t.Type is not null)
+									  .Select(t => $"using {t.Type.ContainingNamespace.ToDisplayString()};")
+									  .ToList();
+
+				var parametersString = "UiContext uiContext";
+
+				var uiContextParameter =
+					SyntaxFactory.Parameter(SyntaxFactory.Identifier("uiContext")
+														 .WithLeadingTrivia(SyntaxFactory.Space))
+								 .WithType(SyntaxFactory.ParseTypeName("UiContext"));
+
+				if (hasCtorArgs)
+				{
+					parametersString += ", ";
+				}
+
+				parametersString =
+					ctor.ParameterList.Parameters.Insert(0, uiContextParameter).ToFullString();
+
+				var usings = string.Join(Environment.NewLine, parameterUsings.Distinct().OrderBy(x => x));
+
+				var code =
+	$$"""
+{{usings}}
+using WalletWasabi.Fluent.Models.UI;
+
+namespace {{namespaceName}};
+
+partial class {{className}}
+{
+    public {{className}}({{parametersString}}){{ctorString}}
+    {
+	    UiContext = uiContext;
+    }
+}
+""";
+
+				var sourceText = SourceText.From(code, Encoding.UTF8);
+				context.AddSource(fileName, sourceText);
+
+				var tree = SyntaxFactory.ParseSyntaxTree(sourceText);
+
+				var newConstructor =
+					tree.GetRoot()
+						.DescendantNodes()
+						.OfType<ConstructorDeclarationSyntax>()
+						.FirstOrDefault();
```

Comment:
It can return null, shouldn't it be checked?

Reply:
@soosr I changed `FirstOrDefault()` to just `First()` because the `newConstructor` can never be null, since it's getting generated right before that code.